### PR TITLE
Add LinearAttention and CausalConvWithState operators (Opset 27)

### PR DESCRIPTION
**Pull request by tadani3**
## Add `LinearAttention` and `CausalConvWithState` operators (opset 27)

Implements the operators specified in RFC [#7767](https://github.com/onnx/onnx/pull/7767) (discussion thread [#7689](https://github.com/onnx/onnx/issues/7689)). Two new standard ops are added at opset 27 to enable efficient representation of linear attention / state-space model layers and the stateful causal 1D convolution that precedes them in modern hybrid LLMs.

### Motivation

A new generation of hybrid LLM architectures — **Qwen3.5 (Gated DeltaNet), Jamba, RWKV-6, Mamba-2, GLA, FalconMamba** — interleaves softmax attention with **linear attention** layers that maintain a fixed-size recurrent state $S \in \mathbb{R}^{d_k \times d_v}$ instead of a growing KV cache.

Today ONNX has **no first-class way to represent these layers**. All ~12 attention-related ops in the standard and ORT contrib domains are softmax-based, and the FlexAttention proposals ([#7494](https://github.com/onnx/onnx/issues/7494), [#6389](https://github.com/onnx/onnx/issues/6389)) extend softmax attention only.

The current workaround — decomposition into ~20 primitive ops via `Scan` — is correct but **10–50× slower** than fused kernels (memory-bound, ~0.18 FLOPs/byte vs. ~90 FLOPs/byte for a fused kernel on A100). For a model like Qwen3.5 where ~75% of layers are Gated DeltaNet, this overhead dominates inference.

These two ops give backends a recognizable target for fused kernels (FLA, causal-conv1d, mamba-ssm) while providing a `Scan`-based reference function body as a universal fallback. See the RFC for full roofline analysis, runtime/dispatch discussion, and architecture coverage.

#### Operators (opset 27, onnx domain)

- **`LinearAttention`** — unified op covering both autoregressive decode (T=1) and chunk-parallel prefill (T>1) through a single interface.
  - 3D packed I/O: `query (B,T,Hq·dk)`, `key/value (B,T,Hkv·dk|dv)`, optional `past_state (B,Hkv,dk,dv)`, optional broadcastable `decay` and `beta`.
  - Attributes: `q_num_heads`, `kv_num_heads`, `update_rule ∈ {"linear","gated","delta","gated_delta"}`, `scale`, `chunk_size`.
  - `stash_type`-style separate type parameter `S` for the recurrent state, allowing fp32 state accumulation with fp16/bf16 activations (LSTM convention).
  - Strict input-combination validation per `update_rule` (per RFC).
  - Supports MHA, GQA, and MQA.

- **`CausalConvWithState`** — stateful causal 1D depthwise conv used as the preprocessing step in Gated DeltaNet / Mamba.
  - Inputs `(B,C,L)` input, `(C,1,k)` weight, optional `bias`, optional `(B,C,k-1)` carry state.
  - Attribute: `activation ∈ {"none","silu","swish"}` (silu/swish aliased).

#### Implementation pieces

| Area | Files |
|---|---|
| C++ schemas + `Scan`-based function bodies | defs.cc (+858), operator_sets.h |
| Python reference implementations | op_linear_attention.py, op_causal_conv_with_state.py, `_op_list.py` |
| Minor `Scan` reference fix | op_scan.py (correct `state_in` naming when loop state vars are present) |
| Generated docs | Operators.md, Changelog.md, TestCoverage.md |
| Node test generators | linear_attention.py, causal_conv_with_state.py |
| Generated node test artifacts | `onnx/backend/test/data/node/test_linear_attention_*`, `test_causal_conv_with_state_*` (model + expanded function-body model + I/O `.pb`) |

#### Backend node tests (with both base and `_expanded` function-body variants)

`LinearAttention`:
- Update rules: `linear`, `gated`, `delta`, `gated_delta`
- Decay shapes: per-head scalar, per-key-dimension
- Beta shapes: per-head, scalar broadcast
- Head configurations: MHA, GQA, MQA
- State handling: explicit zero `past_state`, omitted `past_state`, decode step (T=1), prefill with prior context
- Misc: explicit `scale` override, `T=1` linear path, **fp16 activations with fp32 state**

`CausalConvWithState`:
- Basic; with `bias`; with `past_state`; both together
- Activations: none, `silu`, `swish` alias; with combined past state
- Edge cases: kernel size 1, single-batch single-channel, short input without past state, fp16
- Decode step (single-token streaming)

#### Shape-inference tests (shape_inference_test.py, +691 lines)

`LinearAttention`: basic MHA / GQA / MQA, per-head and per-key-dim decay, with/without past state, dynamic `T`, indivisible head dim, unknown last dims, plus negative tests for rank violations on `query`, `decay`, `past_state`, and for missing/forbidden `decay`/`beta` per `update_rule`.

`CausalConvWithState`: static shapes, all-optional inputs, dynamic length, dynamic kernel, kernel size 1, fp16 type propagation, and negative tests for rank-2 input/weight and zero kernel.

#### Version converter tests

- `automatic_upgrade_test.py` (+130): adapter coverage for both ops across multiple input/attribute configurations.
- `automatic_downgrade_test.py` (+29): confirms downgrade past opset 27 fails as expected (no prior version exists).

### Notes for reviewers

- The RFC document is included for reference at 0009-linear-attention-and-causal-conv.md and tracks all design decisions (3D-only inputs, single op for prefill+decode, separate Q/K/V, `update_rule` as string, `S` type parameter for state precision, etc.). 
- Function bodies are `Scan`-based and serve as both semantic specification and universal fallback. Backends are expected to provide native fused kernels (FLA-style for GPU, AVX-512/NEON for CPU) for performance; the function body guarantees correctness everywhere.
- A small bug in `op_scan.py`'s reference (incorrect `state_in` naming when loop state variables are present) was uncovered while expanding the `LinearAttention` function body and is fixed here.

### CODE REVIEW

File: onnx/defs/operator_sets.h [hunk 1]
```diff
@@ -1399,6 +1399,7 @@ class OpSet_Onnx_ver24 {
 // Forward declarations for ai.onnx version 25
 class ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 25, Cast);
 class ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 25, CastLike);
+class ONNX_OPERATOR_SET_SCHEMA_CLASS_NAME(Onnx, 25, CausalConvWithState);
```
Comment (gramalingam) [inline onnx/defs/operator_sets.h]: I think these ops need to go into (new) opset 27.

File: onnx/defs/nn/defs.cc [hunk 2]
```diff
@@ -3699,4 +3699,861 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+          // Propagate element type from input to both outputs.
+          propagateElemTypeFromInputToOutput(ctx, 0, 0); // input -> output
+          propagateElemTypeFromInputToOutput(ctx, 0, 1); // input -> present_state
+
+          // --- Output 0: same shape as input (B, C, L) ---
+          propagateShapeFromInputToOutput(ctx, 0, 0);
+
+          // --- Input rank validation ---
+          // All tensors are rank-3 for this 1D-only operator.
+          if (hasInputShape(ctx, 0)) {
```
Comment (gramalingam) [inline onnx/defs/nn/defs.cc]: Use [checkInputRank(ctx, 0, 3)](https://github.com/onnx/onnx/blob/fc131ecf091e8a35a07f81906f83c4829ef21140/onnx/defs/shape_inference.h#L827) to simplify this code.

File: onnx/defs/nn/defs.cc [hunk 3]
```diff
@@ -3699,4 +3699,861 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+            auto& input_shape = getInputShape(ctx, 0);
+            if (input_shape.dim_size() != 3) {
+              fail_shape_inference(
+                  "Input tensor must be rank 3 (batch_size, channels, length), got rank ", input_shape.dim_size());
+            }
+          }
+
+          if (hasInputShape(ctx, 1)) {
```
Comment (gramalingam) [inline onnx/defs/nn/defs.cc]: As above ... use checkInputRank utility

File: onnx/test/shape_inference_test.py [hunk 4]
```diff
@@ -2514,6 +2514,697 @@ def test_attention_4d(self) -> None:
             ],
         )
 
+    def test_linear_attention_basic_mha(self) -> None:
+        # update_rule="linear" with no optional inputs: baseline shape/dtype plumbing.
+        # H_q == H_kv == 4, d_k == d_v == 16. Output is 3D packed; state is 4D.
+        graph = self._make_graph(
```
Comment (gramalingam) [inline onnx/test/shape_inference_test.py]: I suggest asking copilot to use `onnx.parser` with a textual/string representation of the model. That will be more readable.

File: onnx/defs/nn/defs.cc [hunk 5]
```diff
@@ -3699,4 +3699,861 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+                .Const1D("Two1D", static_cast<int64_t>(2))
+                .Add("StateExpandShape = Concat <axis = 0> (BatchDim, HkvConst, GroupSize, DkDim, DvDim)")
+                .Add("StateReadShape = Concat <axis = 0> (BatchDim, HqConst, DkDim, DvDim)");
+          }
+
+          // --- Step 5: Scan node with body subgraph ---
+          // The body implements one recurrence step. Per update_rule we
+          // assemble the body from reusable fragments. To keep the body
+          // self-contained (no outer-scope name capture, which several
+          // backends and the ONNX reference Scan do not propagate), we:
```
Comment (gramalingam) [inline onnx/defs/nn/defs.cc]: I don't understand this. Reference to outer-scope variables within the body are perfectly fine (according to ONNX standard). Onnxruntime should support it. I don't know about the ONNX reference Scan. If it doesn't support it, that sounds like a bug.

File: onnx/defs/operator_sets.h (hunk 1 above)
Comment (tadani3) [reply onnx/defs/operator_sets.h]: Addressed in [1bb1d52](https://github.com/onnx/onnx/pull/7950/commits/1bb1d5207bb6de947e496f69e8636f22193df1c1). Moved CausalConv and LinearAttn to Opset 27.

File: onnx/defs/nn/defs.cc (hunk 2 above)
Comment (tadani3) [reply onnx/defs/nn/defs.cc]: Resolved in [5902aa0](https://github.com/onnx/onnx/pull/7950/commits/5902aa02b6cce83e7e6c439b375884ef1b76899b)

File: onnx/defs/nn/defs.cc (hunk 3 above)
Comment (tadani3) [reply onnx/defs/nn/defs.cc]: Resolved in [5902aa0](https://github.com/onnx/onnx/pull/7950/commits/5902aa02b6cce83e7e6c439b375884ef1b76899b). Also clarified for LinearAttention logic in [6faabc9](https://github.com/onnx/onnx/pull/7950/commits/6faabc9e137f3eaf363da90de2dcedadb3c6f8c3)

**Comment by andife** [comment]
@tadani3 can you fix the dco?

File: onnx/test/shape_inference_test.py (hunk 4 above)
Comment (tadani3) [reply onnx/test/shape_inference_test.py]: Resolved in [8c3d1df](https://github.com/onnx/onnx/pull/7950/commits/8c3d1dfa7a109abe094f615655b4052b3f6d5278). Refactored tests using onnx.parser to improve readability.

File: onnx/defs/nn/defs.cc [hunk 6]
```diff
@@ -3697,4 +3699,812 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+          // --- Output 1: present_state shape (B, C, k-1) ---
+          // The carry state stores the last (k-1) input positions for
+          // causal continuity across decode steps.
+          if (hasInputShape(ctx, 0) && hasInputShape(ctx, 1)) {
+            auto& input_shape = getInputShape(ctx, 0);
+            auto& weight_shape = getInputShape(ctx, 1);
+            TensorShapeProto state_shape;
+
+            // Batch and channel dims come from the input.
+            *state_shape.add_dim() = input_shape.dim(0); // B
```
Comment (gramalingam) [inline onnx/defs/nn/defs.cc]: Minor suggestion: please see the bottom of https://github.com/onnx/onnx/blob/main/docs/ShapeInference.md. A useful strategy for implementing such logic is something like below:
```cpp
   Dim B, C;
   unifyInputDim(ctx, 0, 0, B);
   unifyInputDim(ctx, 0, 1, C);
   ...
```
Now, for `k-1`, we may need some of the more complex logic below, since we need to handle the case when K is known  (static constant) differently from dynamic. Even there, you could do
```cpp
   Dim K;
   unifyInputDim(ctx, 1, 2, K);
   if (K.has_dim_value()) {
      ...
```

Two advantages: the utilities take care of safety-checking (if the rank has not been checked). And, you can use unifyInputDim multiple times with same dim B if it recurs in multiple places ... it will check that they are all compatible and pick up the most informative value.

File: onnx/defs/nn/defs.cc [hunk 7]
```diff
@@ -3697,4 +3699,812 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+            "Input tensor with shape (batch_size, channels, length). Channels-first layout.",
+            "T",
+            OpSchema::Single,
+            true,
+            1,
+            OpSchema::Differentiable)
+        .Input(
+            1,
+            "weight",
+            "Depthwise convolution kernel with shape (channels, 1, k) where k is the kernel size.",
```
Comment (gramalingam) [inline onnx/defs/nn/defs.cc]: Curious why there is a 1 in this shape? Is it anticipating a possible generalization in future?

File: onnx/defs/nn/defs.cc [hunk 8]
```diff
@@ -3697,4 +3699,812 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+                // past_state is provided: concat it on the left along axis 2
+                builder.Add("PaddedInput = Concat <axis = 2> (past_state, input)");
+              } else {
+                // No past_state: zero-pad on the left by (k-1)
+                builder.Add("BatchDim = Shape <start = 0, end = 1> (input)")
+                    .Add("ChannelDim = Shape <start = 1, end = 2> (input)")
+                    .Add("ZeroPadShape = Concat <axis = 0> (BatchDim, ChannelDim, Km1)")
+                    .Add("ZeroPad = ConstantOfShape (ZeroPadShape)")
+                    .Add("ZeroPadCast = CastLike (ZeroPad, input)")
+                    .Add("PaddedInput = Concat <axis = 2> (ZeroPadCast, input)");
```
Comment (gramalingam) [inline onnx/defs/nn/defs.cc]: Replace per-node `.Add()` chaining with grouped blocks:

```cpp
// Instead of:
builder.Add("BatchDim = Shape <start = 0, end = 1> (input)")
    .Add("ChannelDim = Shape <start = 1, end = 2> (input)")
    .Add("ZeroPadShape = Concat <axis = 0> (BatchDim, ChannelDim, Km1)")
    ...

// Use:
builder.Add(R"ONNX(
    BatchDim = Shape <start = 0, end = 1> (input)
    ChannelDim = Shape <start = 1, end = 2> (input)
    ZeroPadShape = Concat <axis = 0> (BatchDim, ChannelDim, Km1)
    ZeroPad = ConstantOfShape (ZeroPadShape)
    ZeroPadCast = CastLike (ZeroPad, input)
    PaddedInput = Concat <axis = 2> (ZeroPadCast, input)
)ONNX");
```

Applies to both ops wherever consecutive nodes are unconditionally emitted together.

File: onnx/defs/nn/defs.cc [hunk 9]
```diff
@@ -3697,4 +3699,812 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+          // Shapes are computed once outside the Scan body and captured inside
+          // by name.
+          if (group_size > 1) {
+            builder.Const1D("GroupSize", group_size)
+                .Const1D("Two1D", static_cast<int64_t>(2))
+                .Add("StateExpandShape = Concat <axis = 0> (BatchDim, HkvConst, GroupSize, DkDim, DvDim)")
+                .Add("StateReadShape = Concat <axis = 0> (BatchDim, HqConst, DkDim, DvDim)");
+          }
+
+          // --- Step 5: Scan node with body subgraph ---
```
Comment (gramalingam) [inline onnx/defs/nn/defs.cc]: LinearAttention Step 5 (~120 lines) builds the Scan body via `std::string +=`. Instead, write one complete body per `update_rule`:

```cpp
std::string scan_body;
if (update_rule == "linear") {
    scan_body = R"ONNX(
        q_t_f = Cast <to = 1> (q_t)
        k_t_f = Cast <to = 1> (k_t)
        v_t_f = Cast <to = 1> (v_t)
        KtCol = Unsqueeze (k_t_f, NegOne1D)
        WriteRow = Unsqueeze (v_t_f, NegTwo1D)
        Outer = MatMul (KtCol, WriteRow)
        state_out = Add (state_in, Outer)
        QtRow = Unsqueeze (q_t_f, NegTwo1D)
        ReadOut = MatMul (QtRow, state_out)
        ReadSq = Squeeze (ReadOut, NegTwo1D)
        output_t_f = Mul (ScaleFactor, ReadSq)
        output_t = CastLike (output_t_f, q_t)
    )ONNX";
} else if (update_rule == "gated") { ... }
```

File: onnx/defs/nn/defs.cc [hunk 10]
```diff
@@ -3697,4 +3699,812 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+          // For GQA (group_size > 1) the per-KV-head state must first be
+          // repeat-interleaved across the `group_size` query heads sharing
+          // each KV head. We use the same Unsqueeze + Expand + Reshape
+          // pattern as the standard Attention op. `StateExpandShape` and
+          // `StateReadShape` are captured from the enclosing scope.
+          // The read computes in float32 (using q_t_f and the float32 state),
+          // then casts the per-step output back to T (anchored on the
+          // T-typed q_t) so the Scan accumulates a T-typed output tensor.
+          std::string read_block;
+          if (group_size > 1) {
```
Comment (gramalingam) [inline onnx/defs/nn/defs.cc]: The `group_size > 1` read block (lines 4291–4299) uses C++ string concat despite being static. Match the `group_size == 1` case which already uses a raw literal.

File: onnx/defs/nn/defs.cc (hunk 5 above)
Comment (tadani3) [reply onnx/defs/nn/defs.cc]: Added a fix in https://github.com/onnx/onnx/pull/7964 to solve the outer-scope referencing bug. Marking as resolved.

File: onnx/defs/nn/defs.cc (hunk 6 above)
Comment (gramalingam) [reply onnx/defs/nn/defs.cc]: Please see utilities added in this PR: https://github.com/onnx/onnx/pull/7999

I think we should be able to directly translate the shape declarations (seen in Input(...) above) as follows:
```cpp
   Dim B, C, L, One, K, KMinus1;
   One.set_dim_value(1);
   unifyInputShape(ctx, 0, {B, C, L});
   unifyInputShape(ctx, 1, {C, One, K});
   if (ctx.hasInput(2)) {
      unifyInputShape(ctx, 2, {C});
   }
   if (ctx.hasInput(3)) {
      unifyInputShape(ctx, 3, {B, C, KMinus1});
   }
   // update KMinus1 from K (as below)
   updateOutputShape(ctx, 1, {B, C, KMinus1});
```

For connecting KMinus1 and K (the logic down below), I would suggest defining `operator-` on Dim similar to [this](https://github.com/onnx/onnx/blob/687c4e538d38b46fb2c367ea570414dc5621bccd/onnx/defs/shape_inference.h#L232) and use as
```cpp
   unifyDim( K - 1, KMinus1);
```

Comment (gramalingam) [reply onnx/defs/nn/defs.cc]: With https://github.com/onnx/onnx/pull/8000 you should be able to do what the above comment suggests

File: onnx/defs/nn/defs.cc (hunk 10 above)
Comment (tadani3) [reply onnx/defs/nn/defs.cc]: Resolved in [1dbc321](https://github.com/onnx/onnx/pull/7950/commits/1dbc32104f0273883df0f9a4eeebf92d2b262658) by using a raw literal instead.

Comment (tadani3) [reply onnx/defs/nn/defs.cc]: Fixed in [1dbc321](https://github.com/onnx/onnx/pull/7950/commits/1dbc32104f0273883df0f9a4eeebf92d2b262658) by refactoring to use raw literal.

Comment (tadani3) [reply onnx/defs/nn/defs.cc]: Resolved in [1dbc321](https://github.com/onnx/onnx/pull/7950/commits/1dbc32104f0273883df0f9a4eeebf92d2b262658) by refactoring to use a raw literal.

File: onnx/defs/nn/defs.cc (hunk 8 above)
Comment (tadani3) [reply onnx/defs/nn/defs.cc]: Resolved in [dde3b6c](https://github.com/onnx/onnx/pull/7950/commits/dde3b6caecbefc57cdb6232039807556a682f798) which applied the grouped blocks refactor to both Linear Attention and Causal Conv.

File: onnx/defs/nn/defs.cc (hunk 6 above)
Comment (tadani3) [reply onnx/defs/nn/defs.cc]: Resolved in [7ec3796](https://github.com/onnx/onnx/pull/7950/commits/7ec3796c224efef41413dce6b4c9a3f07fe92a0e)

File: onnx/defs/nn/defs.cc (hunk 9 above)
Comment (tadani3) [reply onnx/defs/nn/defs.cc]: Resolved in [586d926](https://github.com/onnx/onnx/pull/7950/commits/586d926396d4480983a3d9adccbf50f53c6a6817) by adding one complete body literal per update_rule plus two read-tail literals for GQA/non-GQA case.

File: onnx/defs/nn/defs.cc (hunk 7 above)
Comment (tadani3) [reply onnx/defs/nn/defs.cc]: Not for future generalization — the op is fixed as depthwise (no `group` attribute). The `1` keeps the ONNX `Conv` weight layout: `W = (M, C/group, k1, …, kn)`. For depthwise Conv1D, `group = C` so `C/group = 1`, giving `(C, 1, k)`.

Comment (tadani3) [reply onnx/defs/nn/defs.cc]: Will add clarifying comment above

File: onnx/defs/nn/defs.cc [hunk 11]
```diff
@@ -4053,314 +4072,331 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+              // Inputs are reshaped to (B, T, H, D), then transposed so the T axis is
+              // first: (T, B, H, D). The leading T axis is consumed by Scan (which
+              // only supports scan_input_axes=0 in the ONNX reference runtime).
+
+              // Query: (B, T, H_q*d_k) -> (B, T, H_q, d_k) -> (T, B, H_q, d_k)
+              builder.Add(R"ONNX(
+            QShape4D = Concat <axis = 0> (BatchDim, SeqLenDim, HqConst, NegOne1D)
+            QReshaped = Reshape (query, QShape4D)
+            Q4D = Transpose <perm = [1, 0, 2, 3]> (QReshaped)
+          )ONNX");
```
Comment (gramalingam) [inline onnx/defs/nn/defs.cc]: The indentation looks odd. I think that the formatters won't touch this as it is inside a string. We can manually adjust the indentation to what looks good.

Comment (gramalingam) [reply onnx/defs/nn/defs.cc]: So, I would suggest indenting above 4 lines twice (that is, may be 4-blanks each, I am guessing here).

Comment (tadani3) [reply onnx/defs/nn/defs.cc]: Resolved in [a1220fd](https://github.com/onnx/onnx/pull/7950/commits/a1220fde03efed9ff6e70de9fcfea6c609b04df1) by fixing in-lining.

**Review by gramalingam — APPROVED** [review]

File: onnx/defs/nn/defs.cc [hunk 12]
```diff
@@ -3710,4 +3710,817 @@
...
+Weight layout: (channels, 1, k) for depthwise convolution.
+The carry state stores the last (k-1) positions for incremental decode.
+
+The optional activation attribute supports fused SiLU/Swish activation.
+
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    CausalConvWithState,
+    27,
```
[the remark this answers is not in the document]
Comment (tadani3) [reply onnx/defs/nn/defs.cc:3738]: @gramalingam can you dismiss this alert?

**Comment by cbourjau** [comment]
Why are we merging the implementation of an rfc before the rfc (especially given that the rfc has some serious hallucination hallmarks)?
